Pull request statistics

Bullrich opened the pull request:
PR which shows the statistics of all the Pull Requests for a GitHub repository.

Resolves #2 

The code is split into three important sections:
- `github`
  - Interacts with the GitHub api and parses the data.
- `report`
  - Parses the GitHub api data and generates reports with the data split between dates.
- `reporter`
  - Using the parsed reports, it converts them into markdown and returns a `summary` object, that, if it used by the `CLI` it generates a markdown file, and if it is being used by a GitHub action, it writes it to the action's summary.

Bullrich committed:
updated node version to 20

Bullrich committed:
updated gitignore

Bullrich committed:
added class to handle repo data

Bullrich committed:
installed required dependencies

Bullrich committed:
created CLI which generates a markdown document

Bullrich committed:
removed unused method

Bullrich committed:
simplified logic to use only first review

Pagination is not required as we are currently getting the time to the first review

Bullrich committed:
changed names

Bullrich committed:
set code to be runnable as action

Bullrich committed:
added workflow to run itself

So we can use this for testing

Bullrich committed:
fixed log not being a debug log

Bullrich committed:
added more repos to test the action

rzadp commented:
It's statistics, not stadistics :D

rzadp commented on package.json:
```diff
@@ -22,8 +23,13 @@
-    "@actions/github": "^5.1.1",
-    "@octokit/webhooks-types": "^7.3.1"
+    "@actions/github": "^6.0.0",
+    "@octokit/types": "^12.4.0",
+    "@octokit/webhooks-types": "^7.3.1",
```
Seems like those are devDependencies.

Bullrich commented:
> It's statistics, not stadistics :D

Good catch. It's a difficult word 😓 

The PR is currently in Draft, I don't know why it didn't started as a draft (i selected that option).

I moved the status to draft, as I'm currently using the PR to test the action. I'll update it once I polish it (there are **so many linting issues**)

Bullrich committed:
changed action to run with a PAT

This is only for testing

Bullrich committed:
converted metrics to days

As we have more than a thousand of hours in average, it makes sense to use a bigger metric for easier understanding

Bullrich committed:
Fixed time for first review not being an average

Bullrich committed:
added average of additions and deletion

Bullrich committed:
fixed additions to be per month

Being average in total breaks the system

Bullrich committed:
added graphql schema

This makes everything so much cleaner

Bullrich committed:
replaced logic to use graphql result

This simplifies everything into only two methods

It also minimizes the amount of API calls

Bullrich committed:
added calcuations of average reviews per PR

Bullrich committed:
moved graphql file to its own .graphql file

And add a script to copy it after installation

Bullrich committed:
simplified code

All the averages logic is run inside the report tool instead of the printer

Bullrich committed:
linted the project

Bullrich committed:
fixed wrong type

Bullrich committed:
fixed problems with the postinstall script

Now it will skip it in docker

Bullrich committed:
fixed month error with momentjs

Bullrich committed:
replaced heavy repo for lighter

This will reduce the amount of time tha the action takes

rzadp commented on .github/workflows/run-check.yml:
```diff
@@ -0,0 +1,29 @@
+    name: Get repository metrics
+    steps:
+      - uses: paritytech/metrics@pull/stadistics
+        name: Fetch polkadot-fellows/runtimes stadistics
```
```suggestion
        name: Fetch polkadot-fellows/runtimes statistics
```

:D

rzadp replied:
Ah it's the same PR as before, nevermind.

Bullrich replied:
I'm gonna delete that file at the end, it just for testing.

I changed the name of the PR tho, so it should have fixed that

Bullrich committed:
fixed typo

rzadp commented on src/reporter.ts:
```diff
@@ -0,0 +1,126 @@
+): typeof summary => {
+  const prChart = `\`\`\`mermaid
+  pie title Pull Requests for the repository
+  "Open" : ${metrics.open}
+  "Merged" : ${metrics.merged}
+  "Closed" : ${metrics.closed}
```
This is the current number of open PRs, and the total numer of merged/closed PRs of all time, correct?

Bullrich replied:
That's right

rzadp commented on src/util.ts:
```diff
@@ -1,7 +1,86 @@
+};
+
+/**
+ * Calculates how many team a date is repeated in a given month
```
```suggestion
 * Calculates how many times a date is repeated in a given month
```

Bullrich committed:
fixed typo

Co-authored-by: Przemek Rzad <roopert7@gmail.com>

Bullrich committed:
changed action to use github.token

Let's see if it works with this

Bullrich committed:
fixed page number being wrong

Bullrich committed:
changed to test on ink

Bullrich committed:
updated readme

Bullrich committed:
removed unused types

Bullrich committed:
fixed prettier problems

Bullrich committed:
added comment to test action

Bullrich committed:
Added instructions on how to run

Bullrich committed:
moved query to a query folder

rzadp commented on README.md:
```diff
@@ -1,9 +1,62 @@
+    name: Get repository metrics
+    steps:
+      - uses: paritytech/metrics@main
+        name: Fetch polkadot-fellows/runtimes metrics
```
I know this is just an example, but it is named about `polkadot-fellows/runtimes` metrics but gets `polkadot-fellows/metrics` metrics so it's confusing.

Bullrich replied:
I removed it so I don't get confused again

Bullrich committed:
Removed name from metrics

Bullrich committed:
reverted wrong eslint dir to ignore

rzadp commented on src/util.ts:
```diff
@@ -1,7 +1,86 @@
+ * @param getAmount The method to extract such custom value from the object to add to the metrics
+ * @param returnAverage True if the result should be the average for the metrics for a month,
+ * false if it must be the accumulation
+ * @returns Returns an array of objects of type {month: "MM YY", matches: number}
```
Looks like it's actually an array of `[string, number]` arrays, not of `{month, matches}` objects.

```suggestion
 * @returns Returns an array of tuples - a month in the "MM YY" format and a corresponding number of matches.
```

Bullrich replied:
You are right. I changed the type of value and forgot to update that doc. Good catch!

rzadp commented on src/util.ts:
```diff
@@ -1,7 +1,86 @@
+        matches = currentMonthValues.reduce((a, b) => a + b, 0);
+      }
+      // We push the previous match and reset it
+      monthsWithMatches.push([currentMonth.format("MMM YYYY"), matches]);
```
Looks like a mistake. Should it be `"MM YY"` like JSDOC days?

Bullrich replied:
I prefer `MMM YYYY` because it gives _Feb 2024_ vs `MM YY` which gives _02 24_

I feel like it is a bit more clear.

Bullrich committed:
Fixed type changed in docs

Co-authored-by: Przemek Rzad <roopert7@gmail.com>

Bullrich committed:
removed test workflow
